File: ios/Classes/Models/Peer.swift
import Foundation
import CoreBluetooth

/// Connection state of a peer
enum PeerConnectionState: String {
    case discovered
    case connecting
    case connected
    case disconnecting
    case disconnected
}

/// Represents a peer in the BLE mesh network
struct Peer {
    let senderId: String?           // Stable UUID identifier (6-byte compact format)
    let connectionId: String        // CBPeripheral UUID for connections
    let nickname: String
    let rssi: Int
    let lastSeen: Date
    let connectionState: PeerConnectionState
    let hopCount: Int
    let lastForwardTime: Date?
    let isBlocked: Bool
    let peripheral: CBPeripheral?

    init(
        senderId: String? = nil,
        connectionId: String,
        nickname: String,
        rssi: Int = 0,
        lastSeen: Date = Date(),
        connectionState: PeerConnectionState = .discovered,
        hopCount: Int = 0,
        lastForwardTime: Date? = nil,
        isBlocked: Bool = false,
        peripheral: CBPeripheral? = nil
    ) {
        self.senderId = senderId
        self.connectionId = connectionId
        self.nickname = nickname
        self.rssi = rssi
        self.lastSeen = lastSeen
        self.connectionState = connectionState
        self.hopCount = hopCount
        self.lastForwardTime = lastForwardTime
        self.isBlocked = isBlocked
        self.peripheral = peripheral
    }

    /// Convenience property for backward compatibility (returns connectionId)
    var id: String {
        return connectionId
    }

    /// Check if peer can be connected to
    var canConnect: Bool {
        return senderId != nil &&
               connectionState == .discovered &&
               !isBlocked
    }

    /// Convert to a dictionary for sending to Flutter
    func toMap() -> [String: Any] {
        var map: [String: Any] = [
            "connectionId": connectionId,
            "id": connectionId,  // For backward compatibility
            "nickname": nickname,
            "rssi": rssi,
            "lastSeen": Int64(lastSeen.timeIntervalSince1970 * 1000),
            "connectionState": connectionState.rawValue,
            "hopCount": hopCount,
            "isBlocked": isBlocked
        ]

        if let senderId = senderId {
            map["senderId"] = senderId
        }

        if let lastForwardTime = lastForwardTime {
            map["lastForwardTime"] = Int64(lastForwardTime.timeIntervalSince1970 * 1000)
        }

        return map
    }

    /// Create a Peer from a CBPeripheral
    static func fromPeripheral(
        _ peripheral: CBPeripheral,
        rssi: Int = 0,
        senderId: String? = nil
    ) -> Peer {
        return Peer(
            senderId: senderId,
            connectionId: peripheral.identifier.uuidString,
            nickname: peripheral.name ?? "Unknown",
            rssi: rssi,
            connectionState: .discovered,
            peripheral: peripheral
        )
    }
}

File: android/src/main/kotlin/com/ble_mesh/models/Peer.kt
package com.ble_mesh.models

import android.bluetooth.BluetoothDevice

/**
 * Connection state of a peer
 */
enum class PeerConnectionState {
    DISCOVERED,
    CONNECTING,
    CONNECTED,
    DISCONNECTING,
    DISCONNECTED
}

/**
 * Represents a peer in the BLE mesh network
 */
data class Peer(
    val senderId: String? = null,        // Stable UUID identifier (6-byte compact format)
    val connectionId: String,            // MAC address for connections
    val nickname: String,
    val rssi: Int = 0,
    val lastSeen: Long = System.currentTimeMillis(),
    val connectionState: PeerConnectionState = PeerConnectionState.DISCOVERED,
    val hopCount: Int = 0,
    val lastForwardTime: Long? = null,
    val isBlocked: Boolean = false,
    val device: BluetoothDevice? = null  // For connection operations
) {
    /**
     * Convenience property for backward compatibility (returns connectionId)
     */
    val id: String
        get() = connectionId

    /**
     * Check if peer can be connected to
     */
    val canConnect: Boolean
        get() = senderId != null &&
                connectionState == PeerConnectionState.DISCOVERED &&
                !isBlocked

    /**
     * Convert to a map for sending to Flutter
     */
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "senderId" to senderId,
            "connectionId" to connectionId,
            "id" to connectionId,  // For backward compatibility
            "nickname" to nickname,
            "rssi" to rssi,
            "lastSeen" to lastSeen,
            "connectionState" to connectionState.name.lowercase(),
            "hopCount" to hopCount,
            "lastForwardTime" to lastForwardTime,
            "isBlocked" to isBlocked
        )
    }

    companion object {
        /**
         * Create a Peer from a BluetoothDevice
         */
        fun fromDevice(
            device: BluetoothDevice,
            rssi: Int = 0,
            senderId: String? = null
        ): Peer {
            return Peer(
                senderId = senderId,
                connectionId = device.address,
                nickname = device.name ?: "Unknown",
                rssi = rssi,
                device = device,
                connectionState = PeerConnectionState.DISCOVERED
            )
        }
    }
}

